File: LeetCode/2.Add_Two_Numbers/Mario/2.Add_Two_Numbers.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        let dummy = ListNode(-1)
        var p = dummy
        var l1 = l1
        var l2 = l2

        var carry = 0
        while l1 != nil || l2 != nil || carry != 0 {
            let val1 = l1?.val ?? 0
            let val2 = l2?.val ?? 0
            let total = val1 + val2 + carry
            carry = total / 10

            p.next = ListNode(total % 10)
            p = p.next!

            l1 = l1?.next
            l2 = l2?.next
        }

        return dummy.next
    }
}

File: LeetCode/2.Add_Two_Numbers/2.Add_Two_Numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    // iterate l1 and l2 simultaneously, get their vals
    // plus is 0 at first
    // val1 + val2 + plus, if > 9, the plus is 1 else 0
    
    dummy := &ListNode{1, nil}
    node := dummy
    plus := 0

    for l1 != nil || l2 != nil {
        val := plus
        if l1 != nil {
            val += l1.Val
            l1 = l1.Next
        }

        if l2 != nil {
            val += l2.Val
            l2 = l2.Next
        }

        if val > 9 {
            plus = 1
        }else{
            plus = 0
        }

        node.Next = &ListNode{val%10, nil}
        node = node.Next
    }

    if plus > 0 {
        node.Next = &ListNode{plus, nil}
    }

    return dummy.Next
}